VirtualHavruta/vh.py

```diff
@@ -890,3 +1093,330 @@ def get_topic_descriptions(topic_slugs):
         self.logger.info(f"MsgID={msgid}. Final topic descriptions: {final_descriptions}")
         return final_descriptions
 
+    def graph_traversal_retriever(self,
+                                  screen_res: str,
+                                  scripture_query: str,
+                                  enriched_query: str,
+                                  filter_mode_nodes: str | None = None,
+                                  linker_results: list[dict]|None = None,
+                                  semantic_search_results: list[tuple[Document, float]]|None = None,
+                                  msg_id: str = ''):
+        """Find seed chunks based upon linker results or semantic similarity, then traverse the graph to find related chunks in the local neighborhood.
+
+        Details: https://wwwmatthes.in.tum.de/pages/3bd5tm02lihx/Master-s-Thesis-Philippe-Saad, Thesis p. 28f.
+
+        Parameters
+        ----------
+        screen_res
+            The screen_res query, which is used as part of the query to the Sefaria Linker.
+        scripture_query
+            query used to retrieve documents from the vector database
+        enriched_query
+            query enriched with additional context
+        filter_mode_nodes
+            for 'primary' or 'secondary' references, optional
+        linker_results
+            results from the linker
+        semantic_search_results
+            results from semantic search
+        msg_id, optional
+           identifier for slack message, by default ''
+
+        Returns
+        -------
+            list of sorted chunks, sorted by relevance in descending order
+        """
+        # get seed chunks
+        self.logger.info(f"MsgID={msg_id}. [RETRIEVAL] Starting graph_traversal_retriever.")
+        total_token_count = 0
+        collected_chunks = []
+        ranking_scores_collected_chunks = []
+        if linker_results:
+            if semantic_search_results:
+                self.logger.warning(f"MsgID={msg_id}. Both linker results and semantic search results are provided. Using linker results as seeds.")
+            seed_chunks = self.get_linker_seed_chunks(linker_results=linker_results, msg_id=msg_id)
+        elif semantic_search_results:
+            seed_chunks_vector_db = [doc for doc, _ in semantic_search_results]
+            seed_chunks = self.get_chunks_corresponding_to_nodes(seed_chunks_vector_db)
+        else:
+            raise ValueError("One of linker results or semantic search results need to be provided.")
+        # rank seed chunks
+        candidate_chunks, candidate_rankings, token_count = self.rank_documents(
+            seed_chunks,
+            enriched_query=enriched_query,
+            scripture_query=scripture_query
+        )
+        total_token_count += token_count
+
+        n_accepted_chunks = 0
+        while n_accepted_chunks < self.config["database"]["kg"]["max_depth"]:
```

Comment:
I think we should document what max_depth means well -- in this case we follow the uage in the original thesis and that means max number of accepted chunks, right? In that sense it's different from what some people may think to be the depths on a graph...

Reply:
Added an explanatory comment